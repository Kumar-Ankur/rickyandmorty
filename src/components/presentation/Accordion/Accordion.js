import React, { useState, useRef } from "react";
import Chevron from "../Chevron/Chevron";
import "./Accordion.scss";

function Accordion(props) {
  const [setActive, setActiveState] = useState("");
  const [setHeight, setHeightState] = useState("0px");
  const [setRotate, setRotateState] = useState("accordion__icon");
  const content = useRef(null);

  /**
   * @function toggleAccordion
   * @description - toggle accordion and set state (active, height, rotate)
   */
  function toggleAccordion() {
    setActiveState(setActive === "" ? "active" : "");
    setHeightState(
      setActive === "active"
        ? "0px"
        : `${
            content && content.current && content.current.scrollHeight
              ? content.current.scrollHeight
              : 0
          }px`
    );

    setRotateState(
      setActive === "active" ? "accordion__icon" : "accordion__icon rotate"
    );
  }

  return (
    <div className="accordion__section" data-test="component-accordion">
      <button
        className="accordion"
        onClick={toggleAccordion}
        data-test="component-button"
      >
        <p className="accordion__title">{props.title}</p>
        <Chevron className={`${setRotate}`} width={10} fill={"#fff"} />
      </button>
      <div
        data-test="component-accordion-content"
        ref={content}
        style={{ maxHeight: `${setHeight}` }}
        className="accordion__content"
      >
        <div className="accordion__text" data-test="component-accordion-text">
          {props.content}
        </div>
      </div>
    </div>
  );
}

export default Accordion;
