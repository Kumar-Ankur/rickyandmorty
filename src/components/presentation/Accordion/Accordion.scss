/* Style the accordion section */
.accordion__section {
  display: flex;
  flex-direction: column;
  transition: ease-in-out 1s all;
}

@media only screen and (min-device-width: 1024px) and (max-device-width: 1366px) {
  .accordion__section {
    margin-top: 3rem;
  }
}

/* Style the buttons that are used to open and close the accordion panel */
.accordion {
  background: linear-gradient(to right, #3c3b3f, #605c3c);
  color: #444;
  cursor: pointer;
  padding: 18px;
  display: flex;
  align-items: center;
  border: none;
  outline: none;
  transition: ease-in-out 1s all;
}

/* Add a background color to the button if it is clicked on (add the .active class with JS), and when you move the mouse over it (hover) */
.accordion:hover,
.active {
  background-color: #ccc;
  transition: ease-in-out 1s all;
}

/* Style the accordion content title */
.accordion__title {
  font-family: "Open Sans", sans-serif;
  font-weight: 600;
  font-size: 1rem;
  text-transform: uppercase;
  letter-spacing: 1px;
  color: white;
  margin-left: auto;
}

@media only screen and (max-width: 949px) and (min-width: 550px),
  only screen and (max-width: 549px) and (min-width: 400px),
  only screen and (max-width: 399px) and (min-width: 350px),
  only screen and (max-width: 349px) {
  .accordion__title {
    font-size: .8rem;
  }
}

/* Style the accordion chevron icon */
.accordion__icon {
  margin-left: auto;
  transition: transform 0.6s ease;
}

/* Style to rotate icon when state is active */
.rotate {
  transform: rotate(90deg);
}

/* Style the accordion content panel. Note: hidden by default */
.accordion__content {
  overflow: hidden;
  transition: max-height 0.6s ease;
}

/* Style the accordion content text */
.accordion__text {
  font-family: "Open Sans", sans-serif;
  font-weight: 400;
  font-size: 14px;
  padding: 18px;
}
