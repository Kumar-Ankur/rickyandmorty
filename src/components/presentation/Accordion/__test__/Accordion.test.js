import React, { useRef } from "react";
import { shallow } from "enzyme";
import Accordion from "../Accordion";

/**
 * Factory function to create a shallowwrapper for the App component.
 * @function setUp
 * @param { Object } props - Component props specific to this setup
 * @param { Object } state - Initial state for setUp
 * @returns { shallowWrapper }
 *
 */
function setUp(props = {}) {
  const wrapper = shallow(<Accordion {...props} />);
  return wrapper;
}

/**
 * Return shallowWarpper containing node(s) with the given data-test attribute
 * @function findByTestAttr
 * @param {ShallowWrapper} wrapper - Enzyme shallow wrapper to search within
 * @param {string} val - value of data-test attribute for search
 * @returns { shallowWrapper}
 */
function findByTestAttr(wrapper, val) {
  return wrapper.find(`[data-test="${val}"]`);
}

describe("check accordion component", () => {
  let wrapper;
  const setState = jest.fn();
  const useStateSpy = jest.spyOn(React, "useState");
  useStateSpy.mockImplementation((init) => [init, setState]);
  beforeEach(() => {
    wrapper = setUp({ title: "Test", content: "Test" });
  });

  afterEach(() => {
    jest.clearAllMocks();
  })

  test("generate snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("check accordion compoent render without crash", () => {
    const componentAccordion = findByTestAttr(wrapper, "component-accordion");
    expect(componentAccordion.length).toBe(1);
  });

  test("check accordion button render without crash", () => {
    const componentBtn = findByTestAttr(wrapper, "component-button");
    expect(componentBtn.length).toBe(1);
  });

  test("check accordion content render without crash", () => {
    const componentContent = findByTestAttr(
      wrapper,
      "component-accordion-content"
    );
    expect(componentContent.length).toBe(1);
  });

  test("render accordion text witjout crash", () => {
    const accordionText = findByTestAttr(wrapper, "component-accordion-text");
    expect(accordionText.length).toBe(1);
  });

  test("check onclick event on button", () => {
    const componentBtn = findByTestAttr(wrapper, "component-button");
    componentBtn.simulate("click");
    expect(wrapper.props().children[1].ref.current).toBe(null)
  });
});
