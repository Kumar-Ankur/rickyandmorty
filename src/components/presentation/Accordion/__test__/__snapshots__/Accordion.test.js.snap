// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check accordion component generate snapshot 1`] = `ShallowWrapper {}`;
